File: peerDiscoveryServer/src/main/java/it/uniroma3/cheatchat/peerdiscoveryserver/domain/AddressBook.java
package it.uniroma3.cheatchat.peerdiscoveryserver.domain;

import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.OneToMany;
import lombok.Getter;
import lombok.Setter;

import java.util.Set;

@Getter
@Setter
@Entity
public class AddressBook {

    @Id
    private Long id = 1L;

    @OneToMany
    private Set<Peer> peers;

    public void addPeer(Peer peer){
        this.peers.add(peer);
    }

    public void removePeer(Peer peer){
        this.peers.remove(peer);
    }

}

File: address_book/address_book.py
from convos.peer import Peer

class AddressBook:

    peers = []

    def add_peer (self, peer: Peer):
        if peer not in self.peers:
            print(peer)
            self.peers.append(peer)
        else:
            self.peers[self.peers.index(peer)].last_seen = peer.last_seen
    
    def remove_peer (self, peer: Peer):
        if peer in self.peers:
            self.peers.remove(peer)
    
    def get_peers (self):
        return self.peers
    
    def find_peer (self, username: str):
        for peer in self.peers:
            if peer.username == username:
                return peer
        return None
    
    def __str__ (self) -> str:
        string : str = "Address Book\n"
        linecount = 1
        for peer in self.peers:
            string += str(peer) + "\n"
            linecount += 1
        return string